### PR TITLE
Handle query atoms and bonds in SWIG wrappers

### PR DESCRIPTION
**Pull request by jones-gareth**
#### What does this implement/fix? Explain your changes.

Implements the functionality in the Python `rdqueries` module in a new SWIG interface file (`Queries.i`)

Exposes `expandQuery` and `setQuery` for atoms and bonds.  Adds `describeQuery` and `AtomGetSmarts` to atoms.

CSharp and Java tests.

#### Any other comments?

Some other minor additions to the SWIG interface.  A `getBonds` method for ROMol and `getIntProp`, `getBoolProp` and `getDoubleProp` for RDProps.

I had difficulty in exposing the `Match` method in `QueryAtom` and `QueryBond`.  Maybe because the method is virtual or perhaps   it is a reserved word.  Those classes are extended to have a `MatchAtom` and `MatchBond` method.

### CODE REVIEW

**Comment by greglandrum** [comment]
@jones-gareth @ptosco is going to take a first look at this, and then I will do a review too

File: Code/JavaWrappers/Queries.i [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+  void setQuery(const RDKit::QueryAtom *other) {
+	if (other->hasQuery()) {
+	  ($self)->setQuery(other->getQuery()->copy());
+	}
+  }
+}
+
+%extend RDKit::Atom {
+  void ExpandQuery(const RDKit::QueryAtom *other, Queries::CompositeQueryType how=Queries::COMPOSITE_AND, bool maintainOrder=true) {
+  if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:345]:
```suggestion
  PRECONDITION(other, "bad atoms");
  if (other->hasQuery()) {
```

File: Code/JavaWrappers/Queries.i [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+	  const RDKit::QueryAtom::QUERYATOM_QUERY *qry = other->getQuery();
+	  ($self)->expandQuery(qry->copy(), how, maintainOrder);
+	}
+  }
+
+  void setQuery(const RDKit::QueryAtom *other) {
+	if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:352]:
```suggestion
	PRECONDITION(other, "bad atoms");
	if (other->hasQuery()) {
```

File: Code/JavaWrappers/Queries.i [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+	  ($self)->setQuery(other->getQuery()->copy());
+	}
+  }
+}
+
+%extend RDKit::QueryBond {
+  void ExpandQuery(const RDKit::QueryBond *other, Queries::CompositeQueryType how=Queries::COMPOSITE_AND, bool maintainOrder=true) {
+	if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:360]:
```suggestion
	PRECONDITION(other, "bad bonds");
	if (other->hasQuery()) {
```

File: Code/JavaWrappers/Queries.i [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+	  const RDKit::QueryBond::QUERYBOND_QUERY *qry = other->getQuery();
+	  ($self)->expandQuery(qry->copy(), how, maintainOrder);
+	}
+  }
+
+  void SetQuery(const RDKit::QueryBond *other) {
+	if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:367]:
```suggestion
	PRECONDITION(other, "bad bonds");
	if (other->hasQuery()) {
```

File: Code/JavaWrappers/Queries.i [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+	  ($self)->setQuery(other->getQuery()->copy());
+	}
+  }
+}
+
+%extend RDKit::Bond {
+  void ExpandQuery(const RDKit::QueryBond *other, Queries::CompositeQueryType how=Queries::COMPOSITE_AND, bool maintainOrder=true) {
+	if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:375]:
```suggestion
	PRECONDITION(other, "bad bonds");
	if (other->hasQuery()) {
```

File: Code/JavaWrappers/Queries.i [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+	  const RDKit::QueryBond::QUERYBOND_QUERY *qry = other->getQuery();
+	  ($self)->expandQuery(qry->copy(), how, maintainOrder);
+	}
+  }
+
+  void SetQuery(const RDKit::QueryBond *other) {
+	if (other->hasQuery()) {
```
Comment (ptosco) [inline Code/JavaWrappers/Queries.i:382]:
```suggestion
	PRECONDITION(other, "bad bonds");
	if (other->hasQuery()) {
```

File: Code/JavaWrappers/Atom.i [hunk 7]
```diff
@@ -120,4 +140,30 @@
...
+	  // FIX: this should not be necessary
+	  res = RDKit::SmilesWrite::GetAtomSmiles(($self), doKekule, nullptr, allHsExplicit,
+									   isomericSmiles);
+	}
+	return res;
+  }
+
+  // from Python Wrapper
+  std::string describeQuery() {
+	PRECONDITION(($self), "bad atom");
```
Comment (ptosco) [inline Code/JavaWrappers/Atom.i:159]: I don't think you need this `PRECONDITION` inside a method, we do not have it elsewhere.
```suggestion
```

**Review by ptosco — APPROVED** [review]
LGTM; I have left in a few suggestions that you may want to consider.
I also suggest to run `git-clang-format` as there seem to be some trailing spaces here and there.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
Just noticed that the last set of changes seem to have broken the SWIG builds in CI

**Comment by jones-gareth** [comment]
In addition to the broken build, I think that some additional changes are in order.
The extensions in `Queries.i` now creates `unique_ptr`s then returns the released raw pointer.  `%newobject` then wraps that pointer in a `shared_ptr`.  Better to see if we can return the `unique_ptr` with templates for query atoms and bonds and get rid of all the `%newobject` declarations.

**Comment by jones-gareth** [comment]
I'm mistaken - the SWIG `%include <std_unique_ptr.i>` simply calls `release` for templated unique pointers- so all the wrapped functions will need `%newobject` anyway.  Might be better to replace `std::unique_ptr` in @ptosco's changes with `boost::shared_ptr` and we can get rid of all the `%newobject` delarations.

**Comment by jones-gareth** [comment]
I have removed trailing whitespace from `Queries.i`.  Running `clang-format` on the file scrambles it.

**Comment by jones-gareth** [comment]
I have replace the `unique_ptr`s that @ptosco added with `shared_ptr`s.

The '%newobject' statements are no longer required.  I think this looks good.

**Comment by ptosco** [comment]
@jones-gareth I have made a small PR against your branch with a couple additional suggestions to simplify the existing code as well as your addition, then I think the PR can be merged: https://github.com/jones-gareth/rdkit/pull/12

**Comment by jones-gareth** [comment]
@ptosco I have merged your changes to `ROMol.i`.

I also removed `boost::make_iterator` from the `getAtomBonds` and `getAtomNeighbors` in `ROMol.i`

**Comment by ptosco** [comment]
@jones-gareth I have submitted another PR against your branch with a further simplification of your latest changes. This is going to be the last one, I promise!

**Comment by jones-gareth** [comment]
@ptosco I merged your changes. For some reason I thought there were some typing issues that prevented you removing the loop, but I was mistaken.

**Comment by jones-gareth** [comment]
@greglandrum Looks like all the tests ran ok, but there is a timeout.  Is it possible to run them again?

**Comment by greglandrum** [comment]
> @greglandrum Looks like all the tests ran ok, but there is a timeout. Is it possible to run them again?

No need: the tests all passed. The reported failure is one of those weird artifacts that occasionally pops up

File: Code/JavaWrappers/gmwrapper/src-test/org/RDKit/QueryAtomTest.java [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+        String description = queryAtom.describeQuery();
+        expected = "AtomAnd\n  AtomAtomicNum 7 = val\n  less_AtomImplicitHCount 1 <= \n";
+        assertEquals(expected, description);
+
+        String molSmiles = mol.MolToSmiles();
+
+        Atom atom = new Atom(7);
+        atom.setNumExplicitHs(1);
+        mol.replaceAtom(4, atom);
+        mol.clearComputedProps(true);
```
Comment (greglandrum) [inline Code/JavaWrappers/gmwrapper/src-test/org/RDKit/QueryAtomTest.java:65]: This is redundant: `sanitizeMol()` makes the same call: https://github.com/rdkit/rdkit/blob/db952786b670c54695ca4d856218c3b20b8f03d0/Code/GraphMol/MolOps.cpp#L530

File: Code/JavaWrappers/RDProps.i [hunk 9]
```diff
@@ -38,5 +38,8 @@
 
 %include <RDGeneral/RDProps.h>
 
-/* For the time being, assume all properties will be strings */
 %template(setProp)  RDKit::RDProps::setProp<std::string>;
+%template(setIntProp) RDKit::RDProps::setProp<int>;
```
Comment (greglandrum) [inline Code/JavaWrappers/RDProps.i:42]: Nice. I was missing this in my last SWIG-related PR

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
Thanks @jones-gareth !

**Pull request merged**